zutil.c

```diff
@@ -109,3 +109,49 @@ void Z_INTERNAL zng_cfree(void *opaque, void *ptr) {
     Z_UNUSED(opaque);
     zng_free(ptr);
 }
+
+/* Since we support custom memory allocators, some which might not align memory as we expect, 
+ * we have to ask for extra memory and return an aligned pointer. */
+void Z_INTERNAL *zng_calloc_aligned(zng_calloc_cb calloc, void *opaque, unsigned items, unsigned size, unsigned align) {
+    uintptr_t original_ptr;
+    int32_t alloc_size, align_diff;
+    void *ptr;
+
+    /* Allocate enough memory for proper alignment and to store the original memory pointer */
+    alloc_size = sizeof(void *) + (items * size) + align;
+
+    if (calloc) {
+        ptr = calloc(opaque, 1, alloc_size);
+    } else {
+        ptr = zng_calloc(opaque, 1, alloc_size);
+    }
+
+    if (!ptr)
+        return NULL;
+
+    /* Calculate alignment after compensating for storing original pointer */
+    align_diff = ((uintptr_t)ptr + sizeof(void *)) % align;
+
+    /* Store the original pointer for free() */
+    original_ptr = (uintptr_t)ptr + align_diff;
```

Comment:
This doesn't look right, `align_diff` is supposed to include the bytes needed for the original pointer.

Reply:
Calculating the alignment after compensating for the size of the pointer is easier because then we only have to skip ahead those number of bytes and store the original pointer.

Reply:
Yes, but currently you are off by 4/8 bytes in next 4 lines...

Reply:
Should be `original_ptr - sizeof(void *)` to get the location to store the actual "original" pointer returned by allocation callback.

Reply:
I'm not understanding. `original_ptr` is where the original pointer is stored. We want to store it and return the pointer plus the size of the pointer.

I'm trying to compensating for when `align_diff = 0`.

Reply:
If `align_diff` is 0, then you must add `align` and subtract size of `void *`.

Reply:
`align` is guaranteed to always be more than size of `void *`.

Reply:
I've changed it a bit should be resolved now.